.teamcity/src/subprojects/build/core/ProjectCore.kt

```diff
@@ -1,33 +1,60 @@
 package subprojects.build.core
 
 import jetbrains.buildServer.configs.kotlin.v2019_2.*
+import subprojects.*
 
 data class JDKEntry(val name: String, val env: String)
 data class OSEntry(val name: String, val agentString: String, val taskName: String)
 data class JavaScriptEngine(val name: String, val dockerContainer: String)
+data class CoreEntry(val osEntry: OSEntry, val jdkEntry: JDKEntry)
 
 val operatingSystems = listOf(OSEntry("macOS", "Mac OS X", "linkDebugTestMacosX64"), OSEntry("Linux", "Linux", "linkDebugTestLinuxX64"), OSEntry("Windows", "Windows", "linkDebugTestMingwX64"))
 val jdkVersions = listOf(JDKEntry("Java 8", "JDK_18"), JDKEntry("Java 11", "JDK_11"))
-val browsers = listOf(JavaScriptEngine("Chrome/Node.js", "stl5/ktor-test-image:latest"))
+val javaScriptEngines = listOf(JavaScriptEngine("Chrome/Node.js", "stl5/ktor-test-image:latest"))
 
 
 object ProjectCore : Project({
     id("ProjectKtorCore")
     name = "Core"
     description = "Ktor Core Framework"
-    for (os in operatingSystems) {
-        for (jdk in jdkVersions) {
-            buildType(CoreBuild(os, jdk))
-        }
-    }
+
     params {
         param("system.org.gradle.internal.http.connectionTimeout", "120000")
         param("system.org.gradle.internal.http.socketTimeout", "120000")
     }
-    for (os in operatingSystems) {
-        buildType(NativeBuild(os))
+
+    val matrix = operatingSystems.flatMap { os ->
+        jdkVersions.map { jdk -> CoreEntry(os, jdk) }
     }
-    for (browser in browsers) {
-        buildType(JavaScriptBuild(browser))
+
+    val jvm = matrix.map(::CoreBuild)
```

Comment:
It's better to map and concat them all, then do `forEach(::buildType)` and in `dependencies` block setup all dependencies once because you don't use them separately anywhere.

Reply:
Good point. Will fix.